### ISSUE TITLE
Proposed source code to extend Jasper Reporting (in particular for views)

### ISSUE THREAD

**Issue by cirpka** [archimatetool/archi#518]
[ArchiChanges.txt](https://github.com/archimatetool/archi/files/3541594/ArchiChanges.txt)

Phillip and/or Jean-Baptiste,
 
As you have seen in the forum I have tried to ‘improve’ the Jasper reporting capabilities such that you can customize even further what you want to see with a view. Some time ago I downloaded the 4.4 source and worked from there. As my Java expertise is limited I share with you attached the changes so that an expert can take a look at it and decide if this is a viable manner to expose new functionality.
 
I am in the process of making my Jasper Report looking nicely. If there is a kind of ‘reference’ model I will be more than happy to produce ‘my view’ of that report.
 
Meanwhile I learnt that some functionality is implemented in new class since 4.4 (and 4.5) - ViewRelationsDataSource - which returns all relations used in a View which indeed is part of my proposal. The main difference is the ability to sort by type first (instead of name only) which can be handy if layers are to be processed. 

I will also look at the broader issue that you mentioned in your response.

**Comment by Phillipus** [comment]
Thanks for submitting this. It does seem that your proposal is more comprehensive than the new ViewRelationsDataSource and it may be better to implement an extended ViewChildrenDataSource  

Reference: #510

**Comment by Phillipus** [comment]
TBH - I prefer extending ViewChildrenDataSource like this rather than the new  ViewRelationsDataSource class.

This will affect #510 and @NickTononi

**Comment by Phillipus** [comment]
I'll look closer into this over the coming days...

**Comment by Phillipus** [comment]
@cirpka I've created a Git branch "jasper" that contains your changes and some refactoring of the code from me.

Is it possible that you can test this?

**Comment by cirpka** [comment]
Sure. It will probably be tomorrow. Meanwhile I will also produce a Jasper template using the functionality – so that people know what can be achieved.

**Comment by Phillipus** [comment]
I've simplified `ViewChildrenDataSource` even more.

I've changed it so that the general type strings like "elements", "relations", "business" are not case-sensitive but specific class types have to follow the camel case rule like "BusinessActor" to keep things aligned with their usage in jArchi, CSV export and elsewhere.

**Comment by cirpka** [comment]
Phil,

I have not yet checked the branch (I will!) – but I have created on basis of my code a sample report (which of course needs further beautification) to illustrate what now can be achieved. Two more ‘tricks’ I applied

Extension of the FieldDataFactory to really get hold of all model data


        // *** JC Added ***
        if ( "influence_strength".equals(fieldName) && dataElement instanceof IInfluenceRelationship) { //$NON-NLS-1$"
             IInfluenceRelationship relation = (IInfluenceRelationship)dataElement;
             String s = relation.getStrength();
            return StringUtils.isSet(s) ? s : null;
        }

        if ( "access_accesstype".equals(fieldName) && dataElement instanceof IAccessRelationship) { //$NON-NLS-1$"
             IAccessRelationship relation = (IAccessRelationship)dataElement;
             return (Integer)relation.getAccessType();
        }
        return null;


Secondly the ability to have the background/cover/header image next to the model instead with the template such that I can easily use the same template for various customer scenario’s. That requires either some advanced variable setting within the Jasper template

             <variable name="MODEL_FILE" class="java.lang.String" calculation="System">
             <initialValueExpression><![CDATA[((com.archimatetool.jasperreports.data.ArchimateModelDataSource)$P{REPORT_DATA_SOURCE}).getModel().getFile().getAbsolutePath()]]></initialValueExpression>
             </variable>
             <variable name="MODEL_DIRECTORY" class="java.lang.String" calculation="System">
             <initialValueExpression><![CDATA[((com.archimatetool.jasperreports.data.ArchimateModelDataSource)$P{REPORT_DATA_SOURCE}).getModel().getFile().getParent()]]></initialValueExpression>
             </variable>

Or passing two more poperties in the createJasperPrint method like


        String modelFile = fModel.getFile().getAbsolutePath();
        params.put("MODEL_FILE", modelFile); //$NON-NLS-1$

        String modelDirectory = fModel.getFile().getParent();
        params.put("MODEL_DIRECTORY", modelDirectory); //$NON-NLS-1$


Thus far the update …

Jan

**Comment by Phillipus** [comment]
Jan - we can cover those as separate issues once we've nailed this one.

**Comment by cirpka** [comment]
I’m happy to report that everything is working as expected. Should I open other issues for the other proposed extensions? We probably can resolve those in the same branch. Please advise. Please let me also know if I can contribute to documenting some of this Jasper stuff.

Jan

**Comment by Phillipus** [comment]
Hi Jan, yes please for the other issues.

Documentation would be great. In fact it might be good to put them in a wiki page here (I think you should have the right permissions) if you were OK with that?

For this issue, I need to:

1. Squash the commits.
2. Remove the  `ViewRelationsDataSource` class and unit tests.

The "jasper" branch will need to be re-pulled after this.

@NickTononi - do you have time to test this?  We are removing `ViewRelationsDataSource` but this extended class should make your life easier.

**Comment by Phillipus** [comment]
@cirpka Actually, Jan, do you have time for one more test? I just refactored the sort routines and the methods that call them.

**Comment by cirpka** [comment]
Sure will do. Results by end of the day?

**Comment by Phillipus** [comment]
No hurry. Whenever you have time. :-)

**Comment by cirpka** [comment]
Phil, this is unfortunately not working … . Digging into your change … the problem is that the sort by type gets undone due to the subsequent sort by name. The sort by name was a subsort for items of the same type.

**Comment by Phillipus** [comment]
Ah, me trying to be fancy. Latest commit should fix it.

**Comment by Phillipus** [comment]
@cirpka On second thoughts, we may as well continue with the other proposed changes here in this issue. Also, I'll keep the "jasper" branch as it is in GitHub and squash/cherry pick it when we're done.

**Comment by Phillipus** [comment]
### Extension of the FieldDataFactory to really get hold of all model data

So, this is about retrieving Influence Strength and Access Type. Is this a convenience? Can these attributes be retrieved in any other way?

**Comment by cirpka** [comment]
I would not know how. Jasper requires it to be fields …  and this seems to be the easiest way following the relation_target route – and it works. See the sample report I did include. Of course, I’m not the expert … so I like to learn about better ways.

Jan

**Comment by Phillipus** [comment]
#### Extension of the FieldDataFactory to really get hold of all model data 2

I guess setting the file locations is a useful convenience?

**Comment by cirpka** [comment]
That one can be resolved (as shown) in Jasper itself … but it took me a while to figure out. The rational is simple … if you want to use resources (such as the jpg files) that are model specific, it is much easier to have them close to the model (read the same directory). My setup is, that with any project I create specific backgrounds to match the project – and store these together with the archimate model file(s). This way I can produce per project documents.

And in addition to your previous question – is strength (or accesstype) a convenience. The reason I spotted it was an external model where strength was used in stead of name to ‘label’ a relation. Thus when printing relations I missed the text first. With accesstype it is less obvious, that is only the arrow’s direction. But that was the last one to make it complete.

That said … I just thought of the scenario that given an element one would want to know where it is been used. That functionality is of course present in Archi itself. With a document I can envision a scenario where a summary would be enhanced by that reference. And that summary is present – the ability to obtain the various ElementsDataSources (as you rightfully pointed out to me and where the basis for the ViewChildrenDataSource extensions). The “Customizable Report” does include this functionality which I personally find an overkill most of the times due to the missing references. I can take a deeper look into it and come up with a code proposal (likely both in Java and in jrxml). I would be happy to look into it.

Jan

**Comment by Phillipus** [comment]
I see that we append a File.separator here:

`params.put("REPORT_PATH", reportFolder.toString() + File.separator);`

Should we do the same for the model parent path to be consistent?

**Comment by Phillipus** [comment]
I've committed the changes for the attributes and file params for testing.

**Comment by Phillipus** [comment]
I've removed the "jasper" branch, tidied up the commits and repushed the "jasper" branch as a new one. So testers will need to pull again and move their local "jasper" head to the new one.

**Comment by cirpka** [comment]
Thank you very much for all your efforts. I checked out the latest branch and everything works fine!

May I be so bold to share my thoughts with you to ease and complete the Archi model for Jasper. I have given the model some thought and came to the conclusion that we could combine the ElementsDataSource with the ViewChildrenDataSource. In fact, end of last week before the new branch was created I managed to build a version that did this. My rationale is as follows
Within ArchiMate we have a ‘Model’ containing ‘Elements’ and ‘Views’ all having ‘Properties’. You can retrieve ‘Elements’ based on the ‘Model’ returning everything and you can filter them restricting the overall set to those referenced in a ‘View’. Vice versa one can look from the Element’s point of view and restrict the set of all views to those where the element is referenced in.
I tried to verify this and did a few updates – basically adding the ViewChildrenDataSource functionality as new constructors to the ElementsDataSource. While doing so I also had added the ‘reference’ functionality – the ability to get a list of views on which an element is used and a Clone method on the ArchimateModelDataSource. The latter becomes important when working with Jasper parts (e.g. to implement a Table of Contents). That work unfortunately is not yet finished.
In conclusion implementing all this we will have a complete, straight forward model that includes all information. I even tried to model it, and in my desire for simplicity I left out the datasources and the methods each datasource provided to ‘navigate’ to another class. I also wonder if I should call it Element or Concept, but that aside.[cid:image003.jpg@01D56338.61C73020]

If this proposal gets accepted I will update the “Customizable Report” sample to reflect the simplification in the model on the one hand and the addition of new features on the other hand. I do have a premature version available, but I would like to tidy it up and add a few REPORT:… variables to allow end-users to adapt their ‘printed’ reports to their needs.

As a side note. The specific Archi ‘navigation’ methods (such as getPropertiesDataSource) are a challenge in Jasper design environments such as JasperStudio. The preview functionality fails with subreports or subdatasets that iterate over related classes – e.g. properties of an element. That is due to the fact that within the Jasperreport we need to cast to our classes to allow these methods to be called, but casting requires the java files to be accessible in the design environment. I have not yet conquered this challenge, which even gets complicated as we also need to have a model with some real data to render.

Kind regards

Jan

**Comment by Phillipus** [comment]
I haven't had time to think this through yet, but just to quickly say that we must be careful not to break the API for existing templates that users will have made.

**Comment by Phillipus** [comment]
@cirpka It might be possible then to keep the existing code API and delegate the methods of  ViewChildrenDataSource to the ElementsDataSource so that we can still support existing templates?

Do you have any code to share?

**Comment by cirpka** [comment]
Didn’t you get my ZIP? I’ve done it all (as a non Java programmer – getting better though😉)
Yes, redirection should be possible. The signatures only had to change with respect to casting. I guess my code is not yet ready for this.

**Comment by Phillipus** [comment]
> Didn’t you get my ZIP?

No, was it an attachment? If you are answering these posts by email and not in GitHub itself it might not work.

**Comment by cirpka** [comment]
I only know to work by mail for this. Yep, it was an attachment. I’ll upload it and send you a link … just a couple of minutes – internet in train is slow ☹

**Comment by cirpka** [comment]
https://1drv.ms/u/s!At0_pye8m74Xhdg9w0n1Rwo9yLALLQ?e=Ozh56O

**Comment by Phillipus** [comment]
Regarding the clone method. Would a rewind() or reset() method work instead where the pointer is reset to the beginning?

**Comment by cirpka** [comment]
Fully agree! (Mail got into spam, hence the later answer). Using @depreciated does not affect existing reports. The ViewChildElementsDataSource was untouched. Now that I think of it, obtaining it from the ModelDataSource may got changed … maybe we can superclass it with ElementsDataSource (or vice versa) such that the casting does not break. Begin next week I can take a look into this. Let me know.

**Comment by cirpka** [comment]
Consistency is always preferred. As it is new, please do!

**Comment by cirpka** [comment]
I don’t know. I noticed that being called from a new part the ‘pointer’ was always at the end. Thus the core question is – are parts being processed in parallel. Guess that we need to try. Please give me until begin next week.  OK?

**Comment by Phillipus** [comment]
Where do you use `clone()`? 

Each data source class implements `JRRewindableDataSource` which has a `moveFirst()` method. However, `ArchimateModelDataSource` does not implement it so it could simply be a case of adding it. Then you could call `modelDataSource.moveFirst()` instead of `modelDataSource.clone()`

I'll take your code, merge it, tweak it, refactor it and push into the "jasper" branch over the coming days.

**Comment by cirpka** [comment]
I call clone() passing the ModelDataSource to the part within the JasperReport. As said, I’ll give it a try – and with your heads up it should be a fairly trivial check (if I had no other work to do …).

I was indeed hoping that a master is looking at my code … you did a great job beautifying the other. Meanwhile I will continue to enhance the “Configurable Report”, so that with a release people can actually see the improvements.

**Comment by cirpka** [comment]
Phil,

I just made ArchimateModelDataSource rewindable and implemented moveFirst(). The problem though is, that I need to pass as DataSource to the Jasper part. Clone does return the DataSource, moveFirst is void. Of course, I could evaluate a moveFirst() statement first, but that does not feel only weird, it also may cause some issues in timing due to the different options for evaluation that Jasper knows of (and me a little by now too 😉). I’m not sure if making ArchimateModelDataSource rewindable with it’s one node (holding underneath the model’s properties and the access methods to the ‘real’ stuff) is really adding value, but as the implementation is trivial you may just add it. But I fear that it does not solve the Jasper parts challenge.

Jan

**Comment by Phillipus** [comment]
Perhaps another option is to add a reset() method that returns "this"?

**Comment by cirpka** [comment]
I tried this, technically it works, but the rendered results are wrong. The first part, the title, is empty. Using clone() instead everything is rendered fine. It's probably what I anticipated, the parallel processing.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Phillipus** [comment]
I've tidied, refactored and removes the deprecated code. I did some quick testing and it seems backward compatible. I removed your two experimental methods for now so it compiles.

Pushed to the "jasper" branch,

**Comment by cirpka** [comment]
Thanks Phil,

I just took a look at it and noticed that with experimental methods you likely referred to the ‘cross referencing’ methods getReferencedViews(). I used this method to allow e.g. in the overview of all elements of a specific type to show in which view it is been used. Is that of any value? I guess yes, as having e.g. all business elements at the end one could be interested to see where these elements are actually used.

I am still looking for a good model file with just a few views to demonstrate all the proposed capabilities. If you have one, feel free to share it with me so that I can produce a full example report. All my reports are quite customer specific or just ‘testing’.

Jan

**Comment by Phillipus** [comment]
> I just took a look at it and noticed that with experimental methods you likely referred to the ‘cross referencing’ methods getReferencedViews(). I used this method to allow e.g. in the overview of all elements of a specific type to show in which view it is been used. Is that of any value? I guess yes, as having e.g. all business elements at the end one could be interested to see where these elements are actually used.

I've just implemented it if you'd like to test.

I am still looking for a good model file with just a few views to demonstrate all the proposed capabilities. If you have one, feel free to share it with me so that I can produce a full example report. All my reports are quite customer specific or just ‘testing’.

> I'm that last person to ask. :-) My models are all "Hello World" level. ;-)

**Comment by cirpka** [comment]
Phil,

Sorry for anwering late – the good news is that everything works as expected. What I want to do is to provide a newer version of the Customizable Report to illustrate what now can be done with the extensions that are made. That work still needs some finishing up, so please allow for some time.

Jan

**Comment by Phillipus** [comment]
Hi Jan, no problem about answering late. Glad it all works. 

We should get @jbsarrodie in the loop to check any changes to the Customizable Report.

**Comment by cirpka** [comment]
That’s a good idea, though I would propose to offer a second “Customizable Report” next to the current implementation. Not everyone may like what I did 😉.

I will give it a thought for a name – “Advanced Customizable Report” would be a claim I would not dare to think about. And Customizable Report Version 2 also implies “better” whereas I believe it is primarily different. As said, I’ll give it a thought.

Jan

**Comment by Phillipus** [comment]
One thing I think we could change is lumping Technology and Physical classes together in `ArchimateModelDataSource#getClasses()`:

```
case ELEMENTS_TECHNOLOGY:
   eClasses.addAll(Arrays.asList(ArchimateModelUtils.getTechnologyClasses()));
   eClasses.addAll(Arrays.asList(ArchimateModelUtils.getPhysicalClasses()));
   break;
```

We could separate these into "technology" and "physical" class strings and if they are needed together the query string would be "technology|physical"

But I'm not sure if we need to do that or not?

**Comment by Phillipus** [comment]
New commit - added size() method to all data sources, cast return values and added more unit tests.

**Comment by Phillipus** [comment]
Hi Jan,

what is your conclusion on the new API? Is it all OK?

**Comment by cirpka** [comment]
Yes. It's working fine. I still like to contribute with a new Jasperreport template, but I'll be today and next week on a short break. If you want to release without template it's ok. You certainly can merge it to into the main branch.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Phillipus** [comment]
What do you think about my proposal in the comment above about separating Physical and Technology? - https://github.com/archimatetool/archi/issues/518#issuecomment-531457659

I'm aiming to get this into Archi 4.6, which will be in beta version in a week or so, so no hurry on the template

**Comment by Phillipus** [comment]
> We could separate these into "technology" and "physical" class strings and if they are needed together the query string would be "technology|physical"

I've pushed a commit that does this. Separating them gives you more options

**Comment by jbsarrodie** [comment]
> We should get @jbsarrodie in the loop to check any changes to the Customizable Report.

Hi. FWIW, I follow this thread and will test this at some point (maybe next week)

**Comment by Phillipus** [comment]
I've rebased the "jasper" branch to the latest master branch.

**Comment by Phillipus** [comment]
@cirpka Just a heads up that the version of Eclipse target is now 4.13. So the archi.target file will need to be reloaded and any runtime target configurations re-done. See https://github.com/archimatetool/archi/issues/531

**Comment by Phillipus** [comment]
I'm assuming that the code changes in the "jasper" branch are non-breaking and can be merged into master soon.

**Comment by Phillipus** [comment]
Merged "jasper" branch into master.

**Comment by Phillipus** [comment]
Close this one now that 4.6 is released.
